Issue #8 - Generate Docker Compose file from config

NoxHarmonium opened the pull request:
This is where I got to this afternoon. It isn't finished but Justin suggested to post this PR for feedback.

I have changed the way different tasks are called by using the "command" feature of scopt rather than using switches. I thought this made sense when there are multiple tasks. Let me know what you think.

Still to do:

- [X] Labels (`source.path` and `project.artefact` on components)
- [X] Mapping the edge to a Nginx container with an exposed port
- [X] Environment variables
- [X] Code tidy up
- [X] More unit tests
- [X] Conversion to cats

Blocked:
- [ ] Explicit network names 
- [ ] Integration with new plugin system

Commit:
[8] Created boilerplate for generate task

- Refactored tasks into classes
- Use scopt "command" constructs to seperate tasks

Commit:
[8] First implementation of Generate function

Commit:
[8] Fix tests broken by change to interface

Commit:
[8] Add initial test for generate task

ewheath commented on src/main/scala/au/com/agiledigital/toolform/app/ToolFormApp.scala:
```diff
@@ -24,17 +27,44 @@ object ToolFormApp extends App {
   private def parseCommandLineArgs(args: Array[String]): Either[ToolFormError, ToolFormConfiguration] = {
     val parser = new scopt.OptionParser[ToolFormConfiguration]("toolform") {
-      head("toolform", "0.1")
+      head(PackageInfo.getPackageName, PackageInfo.getPackageVersion)
```
I had the same thought here and used sbt-buildinfo to generate this.  Sorry about the conflict.  Should easy to fix up...

NoxHarmonium replied:
sbt-buildinfo is probably a better way to do it because it is compile time. This reflection method doesn't seem to work in test environments either. I just didn't know if I should be pulling in dependencies or not.

ewheath commented:
I don't have a strong preference for command style vs -/-- style in general... this looks cool as you've done it.
Before too long we might think about types of commands (both of inspect and generate want to accept the same style of input file, for example).

In #7 I'm trying to make the commands/tasks more dynamic so that they can be loaded as plugins. 
That work will effect:
- the command line parsing definitions,
- ToolFormAppMode
- and command exec (Task trait).

If I give you something like:
```
trait ToolFormPlugin {
  def commandName: String
  def commandVersion: String
  def commandLineArgs: OptionDef[Unit, ToolFormConfiguration] // Or a custom return type?
  def createTask(toolConfig: ToolFormConfiguration): Task
}

trait Task {
  def run(project: Project): Either[ToolFormError, String] // TODO: Might want multiple files/outputs?
}
```
do you think we could make it work?

NoxHarmonium commented:
Yeah that makes sense to me.

For `commandLineArgs` I think a custom type might be better because different tasks have different requirements that might not be shared. E.g. Generate needs to know what format to output (Docker or Kubernetes) but the other tasks don't need that. Also we would need to update the core project if we add a new task that needs options we haven't thought of. 

Some of the outputs will be multiple files. For example, Kubernetes needs two yaml files for each service. Maybe a task could handle the file output itself and the string returned by the `run` method is just to print to stdout for logging purposes? Although we might want the logging to happen in realtime, rather than just all at the end of the task.

NoxHarmonium commented:
It also depends on how many tasks we are going to have. If most of the tasks are just taking the project information and doing something with it and Generate is an exception then maybe we don't have to worry about a custom return type in `commandLineArgs`.

ewheath commented:
@NoxHarmonium I was riffing off your changes a bit and have a temp PR up here: https://github.com/agiledigital/toolform/pull/11 - take a look if you have time and we could chat on slack a bit later in the day maybe... nothing set in stone here just ideas on my end.  It's tricky to get the plugin interface lined up so that we can have super dynamic command line and exec flow.

Commit:
[8] Add labels to Docker Compose output

Commit:
[8] Implement edge handling

Commit:
[8] Remove dev-resources.conf

- After talking with JS we decided to remove dev-resources.conf and to
put the environment variables and ports into project.conf
- Component and Project can now have "exposed-ports" and "environment"
properties that are passed through to the output.
- Resources now have an "image" property to specify where the image is
located explicitly
- I added sorted maps to some of the models because the ordering was not
stable when iterating project components. This is only important because
for testing purposes.

Commit:
[8] Some cleanup

- The scala auto formatter has done another pass
- Converted more enums to enumeratum
- Changed method name to identity

Commit:
[8] Make inspect ordering stable

Commit:
[8] Auto formatter

NoxHarmonium commented:
I've implemented labels, generation of the nginx service from a subedge and the ability to provide environment variables and exposed ports for any service (component or resource).

Now that we can specify environment variables and ports directly from `project.conf` we no longer need `dev-resources.conf` so it has been removed.

The generation code is getting pretty ugly. At a minimum it will need to be split up into seperate files but ideally I should rework it to be more readable.

The only easy way to set a custom network name seems to be through a [switch passed to compose or an environment variable](https://docs.docker.com/compose/networking/). Setting it through the compose file is different for each network driver. Apparently compose uses the "bridge" network on a developer machine and the "overlay" network on swarm so we would have to generate a different config for a developer machine or for deployment.

I'll keep on working on it tomorrow.

NoxHarmonium commented:
Also we should be able to test it out on a real project now. I could see how it goes with SETI as I am already familiar with it's setup.

Commit:
[8] Use a scalaz state monad to clean up code

Commit:
[8] Remove unused option

- Removing the catch all from the switch between output types makes the
compiler actually check if every case is handled. This means if somebody
adds a new output type the compiler will tell you if it is not handled.

Commit:
[8] Major refactor to clean up generation code

- Created a package for docker compose file generation and split up
GenerateTask
- Merge together redundant functions for readability
- Grouped indented functions under nested 'for' statements so you can
tell by the code whether something is indented or not

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/SubEdgeDef.scala:
```diff
@@ -0,0 +1,43 @@
+    * @return a iterable of SubEdgeDef objects.
+    */
+  def subEdgeDefsFromProject(project: Project): Iterable[SubEdgeDef] =
+    project.topology.sortedEdges.flatMap((edgePair) => {
```
Prefer:
```
flatMap { case(edgeName, edge) =>

}
```

NoxHarmonium replied:
Converted to a `for` comprehension as discussed in person

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/SubEdgeDef.scala:
```diff
@@ -0,0 +1,43 @@
+  def subEdgeDefsFromProject(project: Project): Iterable[SubEdgeDef] =
+    project.topology.sortedEdges.flatMap((edgePair) => {
+      val (edgeName, edge) = edgePair
+      edge.sortedSubEdges.map((subEdgePair) => {
```
Ditto.

NoxHarmonium replied:
Converted to a `for` comprehension as discussed in person

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/GenerateTask.scala:
```diff
@@ -0,0 +1,37 @@
+object GenerateTaskOutputType extends Enum[GenerateTaskOutputType] {
+  val values = findValues
+
+  case object dockerComposeV3 extends GenerateTaskOutputType
```
Uppercase first letter of enum entries.

NoxHarmonium replied:
Done for all enums in the project

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/YamlWriter.scala:
```diff
@@ -0,0 +1,38 @@
+trait YamlWriter {
+
+  type Result[A] = State[WriterContext, A]
+  val indentSize = 2
```
`private val`?

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/YamlWriter.scala:
```diff
@@ -0,0 +1,38 @@
+}
+
+/**
+  * An object that is used provide context to the generator while it is writing.
```
"Provides context to the generator..."

NoxHarmonium replied:
Fixed

Commit:
[8] Define public member type explicitly

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/DockerFormatting.scala:
```diff
@@ -0,0 +1,47 @@
+    normaliseImageName(s"${projectId}/${component.id}")
+
+  def resourceImageName(resource: Resource): String =
+    normaliseImageName(s"${resource.image}")
```
Why is this using `s"..."` instead of just `resource.image`?

NoxHarmonium replied:
Good point. All examples of this have been removed.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+  def runDockerComposeV3(toolFormConfiguration: ToolFormConfiguration, project: Project): Either[ToolFormError, String] = {
+    val outFile = toolFormConfiguration.generateTaskConfiguration.out
+    if (outFile.isDirectory) {
+      return Left(ToolFormError("Output path is a directory. Docker Compose V3 output requires a single file as an output."))
```
This should be checked when parsing the options for the command. Kubernetes generation should parse options differently.

NoxHarmonium replied:
I have removed this check from `GenerateDockerComposeV3`

I would add the validation to the command line parsing but I've heard that is changing soon anyway so I'll leave it for now.

Commit:
[8] Add tests for port/environment writing

- Fixed a bug where an empty list/map would generate invalid yaml
- Changed WriterContext to use a generic Writer class so any writer can
be passed in. This makes testing a lot easier

Commit:
[8] Convert method to a for comprehension

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+        _ <- write(s"# Generated by ${BuildInfo.name} (${BuildInfo.version})")
+        _ <- write("version: '3'")
+        _ <- write("services:")
+        i <- indent()
```
```
i <- indent()
_ <- ...
_ <- resetIndent(i)
```

occurs frequently. Consider how to express as something like:
```
_ <- indented {
  writeEdges(...),
  writeComponents(...),
  writeResources(...) 
}
```

NoxHarmonium replied:
I have used a construct that looks like this

```
_ <- indented { 
  for {
    _ <- writeEdges(...)
    _ <- writeComponents(...),
    _ <- writeResources(...) 
  } 
}
```

Providing the method calls as parameters to a method as shown in your example could be done and looks a bit neater but would require some plumbing to chain the monads inside the indented function which could add complexity. This way is simpler to implement and maintain and still has the advantage of making the intention clear. 

What do you think?

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/DockerFormatting.scala:
```diff
@@ -0,0 +1,47 @@
+
+  def subEdgePortDefinition(subEdgeDef: SubEdgeDef): String =
+    SubEdgeType.withNameInsensitive(subEdgeDef.subEdge.edgeType) match {
+      case SubEdgeType.http  => "- \"80:80\""
```
Should this delegate to `formatPort(80)`?

NoxHarmonium replied:
Good pickup! This has been done

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+      i <- indent()
+      _ <- for {
+
+        _ <- write(s"source.path: \042${component.path}\042")
```
`\"`

NoxHarmonium replied:
There is a bug in Scala where it doesn't handle `\"` with string interpolation!

This was the workaround I used: https://stackoverflow.com/a/39457924

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  private def writeEnvironmentVariables(service: Service) = service.environment match {
+    case Some(environmentEntries) =>
```
`environment` should be a `Seq[...]` not `Option[Seq[...]]`, I think that there were some issues making this true earlier. Should fix now so that it isn't bleeding into the generators.

NoxHarmonium replied:
It seems like the parser throws an exception if the key isn't defined. I think we would have to make it mandatory to define empty environment/port entries.

NoxHarmonium replied:
Oh never mind you can specify default values on case classes

NoxHarmonium replied:
Fixed by specifying a default value for environment and exposed ports in the case class.

This has now be simplified to an `if` statement checking if the list/map is empty or not.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  private def writePorts(service: Service) =
+    service.exposedPorts match {
```
Ditto.

NoxHarmonium replied:
Fixed by specifying a default value for environment and exposed ports in the case class.

This has now be simplified to an `if` statement checking if the list/map is empty or not.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+    try {
+
+      val writeFile = for {
+        _ <- write(s"# Generated by ${BuildInfo.name} (${BuildInfo.version})")
```
Lets include generation data and source file location in the preamble.

[EDIT - a word]

NoxHarmonium replied:
> generation data

What kind of data do you mean specifically?

NoxHarmonium replied:
Source file location and date have been added to the preamble

dspasojevic commented on src/test/scala/au/com/agiledigital/toolform/app/GenerateTaskTest.scala:
```diff
@@ -12,7 +12,7 @@ class GenerateTaskTest extends FlatSpec with Matchers {
   val testFile = pathToFile("/test_project/environment.conf")
 
   def pathToFile(pathToFile: String): File = {
```
Rename file - it is a test for the Docker V3 generation.

I'd consider having folders of input files and expected output files that are picked up by the test automatically when a new case is added.

NoxHarmonium replied:
The tests now scans the `src/test/resources/testprojects` directory and for each directory it looks for an `environment.conf` file as the input and an `expected.yaml` as the output.

It ignores comments because they will change according to version/environment/time and date.

I put a README in the directory to explain this for future developers.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+import scalaz.Scalaz._
+import scalaz._
+
+class GenerateDockerComposeV3() extends YamlWriter {
```
Add comments to this and `runDockerComposeV3`. Top-level types and the method that serves as the main entry point into an instance should be well documented.

NoxHarmonium replied:
Done

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala:
```diff
@@ -0,0 +1,175 @@
+
+  // TODO: Surely Scalaz has a built in identity function that could replace this.
+  // Maybe it could be replaced with a different construct altogether?
+  private val identity = State[WriterContext, Unit] { context =>
```
Move to companion object.

NoxHarmonium replied:
I have moved everything in this file to a companion object because the whole thing is stateless

dspasojevic reviewed (commented):
Looking nice. Almost there I think.

Commit:
[8] Write some more tests

Commit:
[8] Remove redundant braces

Commit:
[8] Remove redundant code

Commit:
[8] Add constuct for indenting blocks

Commit:
[8] Make member private

Commit:
[8] Move pure functions to companion object

Commit:
[8] Add more information to preamble

Commit:
[8] Remove check that should be done somewhere else

Commit:
[8] Capitalise enum values

Commit:
[8] Remove unneccessary Option wrapper

Commit:
[8] Move test to better location

Commit:
[8] Apply code style

Commit:
[8] Run tests using a folder of test cases

Iterating over a folder of test cases is much easier than writing
manual tests for each scenario and putting the result in a triple quoted
string.

Commit:
[8] Fix issue with components not parsing properly

- The previous mapping method seemed to break the parsing of kebab case
- This was not discovered earlier because the only fields with spaces in
them ('componentGroup' and 'exposedPorts') did not have test coverage
- Renamed 'optionalId' to 'idOverride' to make intention a bit more
clear

Commit:
[8] Add test for naming

Commit:
[8] Add test for edge port mapping

Commit:
[8] Remove TODO

Commit:
[8] Make GenerateDockerComposeV3 class an object

Every function is pure so it doesn't make sense to be a class

NoxHarmonium commented:
I will convert from Scalaz to Cats (https://github.com/typelevel/cats) as discussed in person because the new CLI library already pulls it in as a dependency and it will remove redundancy.

Commit:
[8] Convert from scalaz to cats

- Cats is a dependency for the new CLI library we will be bringing in
therefore it makes sense to use cats so we don't have two functional
programming libraries in the project.

Commit:
[8] Homogenise field mapping style

Commit:
[8] Minor cleanup

Commit:
[8] Fix name of file to match class

NoxHarmonium commented:
@dspasojevic Feedback has been addressed - ready for review

Commit:
[8] Remove trait association that is no longer used

Commit:
[8] Document the Service trait
